psh/date/date.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * Phoenix-RTOS
+ *
+ * date - print or set the system date and time
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include "../psh.h"
+
+
+enum mode { mode_skip,
+	mode_help,
+	mode_print,
+	mode_set,
+	mode_parse };
+
+
+static void psh_date_info(void)
+{
+	printf("print/set the system date and time");
+}
+
+
+static void psh_date_help(const char *prog)
+{
+	printf(
+		"Usage: %s [-h] [-s EPOCH] [-d @EPOCH] [+FORMAT]\n"
+		"  -h:  shows this help message\n"
+		"  -s:  set system time described by EPOCH (POSIX time format)\n"
+		"  -d:  display time described by EPOCH (POSIX time format)\n"
+		"  FORMAT: string with POSIX date formatting characters\n"
+		"NOTE: FORMAT string not supported by options: '-s', '-d'\n",
+		prog);
+}
+
+
+static int psh_date_print(const struct timeval *tv, const char *format)
+{
+	struct tm *tinfo;
+	const char *defformat = "+%a, %d %b %y %H:%M:%S";
+	char buf[64];
+
+	if (format == NULL) {
+		format = defformat;
+	}
+	else if (format[0] != '+') {
+		fprintf(stderr, "date: invalid format '%s'\n", format);
+		return 1;
+	}
+
+	tinfo = localtime(&(tv->tv_sec));
+
+	/* passing `&format[1]` for omission of '+' at the beginning */
+	if (strftime(buf, sizeof(buf), &format[1], tinfo) == 0) {
+		fprintf(stderr, "date: '%s' expansion too long\n", format);
+		return 1;
+	}
+	printf("%s\n", buf);
+
+	return 0;
+}
+
+
+static int psh_date_get(const char *format)
+{
+	struct timeval tv;
+
+	time(&(tv.tv_sec));
+	return psh_date_print(&tv, format);
+}
+
+
+static int psh_date_convert(char *timestring, const char *format, struct timeval *tv)
+{
+	char *end;
+
+	/* FIXME: no strptime() implementation in libphoenix */
+	if (format != NULL) {
+		fprintf(stderr, "date: chosen option does not support FORMAT\n");
+		return 1;
+	}
+
+	if (timestring == NULL) {
+		return 1;
+	}
+
+	tv->tv_usec = 0;
+	tv->tv_sec = strtoull(timestring, &end, 10);
+	if ((end == timestring) || (*end != '\0')) {
+		fprintf(stderr, "Unrecognized argument: %s\n", timestring);
+		return 1;
+	}
+	return 0;
+}
+
+
+static int psh_date_parse(char *timestring, const char *format)
+{
+	struct timeval tv;
+
+	if (timestring != NULL && timestring[0] != '@') {
```

Comment:
Well... yea. `timestring == NULL` is supposed to be check in `psh_date_convert`, however couple of lines later `timestring[1] is passed so it is invalid. 

It is though impossible to pass `NULL` from command line as it would require to pass `NULL` string from `getopt()`. Rearranged checking so that it is done inside `psh_date_convert` once and no access is done before that check. Just to be safe inside `date`. Thanks for pointing it out!